playground/backend/internal/storage/cloud_storage.go

```diff
@@ -0,0 +1,218 @@
+type SdkCategories map[string]ExamplesInfo
+type Examples map[string]SdkCategories
+
+// CloudStorage represents working tools for getting compiled and
+// run beam examples from Google Cloud Storage. It is required that
+// the bucket where examples are stored would be public,
+// and it has a concrete structure of directories, namely:
+// /SDK_JAVA
+// ---- /Category1
+// --------/Example1
+// --------/Example2
+// ----  ...
+// /SDK_GO
+// ---- /Category2
+// --------/Example1
+// --------/Example2
+// ----  ...
+// ...
+type CloudStorage struct {
+}
+
+func New() *CloudStorage {
+	return &CloudStorage{}
+}
+
+// GetExample returns the source code of the example
+func (cd *CloudStorage) GetExample(ctx context.Context, examplePath string) (*string, error) {
+	extension := getFileExtensionBySdk(examplePath)
+	data, err := cd.getFileFromStorage(ctx, examplePath, extension)
+	if err != nil {
+		return nil, err
+	}
+	result := string(data)
+	return &result, nil
+}
+
+// GetExampleOutput returns the run output of the example
+func (cd *CloudStorage) GetExampleOutput(ctx context.Context, examplePath string) (*string, error) {
+	data, err := cd.getFileFromStorage(ctx, examplePath, OutputExtension)
+	if err != nil {
+		return nil, err
+	}
+	result := string(data)
+	return &result, nil
+}
+
+// GetListOfExamples returns the list of stored example at cloud storage bucket
+func (cd *CloudStorage) GetListOfExamples(ctx context.Context, sdk pb.Sdk, category string) (*Examples, error) {
+	client, err := storage.NewClient(ctx, option.WithoutAuthentication())
+	if err != nil {
+		return nil, fmt.Errorf("storage.NewClient: %v", err)
+	}
+	defer client.Close()
+
+	ctx, cancel := context.WithTimeout(ctx, time.Second*Timeout)
+	defer cancel()
+
+	bucket := client.Bucket(BucketName)
+	prefix := sdk.String()
+	if sdk == pb.Sdk_SDK_UNSPECIFIED {
+		prefix = ""
+	}
+	examples := make(Examples, 0)
+	it := bucket.Objects(ctx, &storage.Query{
+		Prefix: filepath.Join(prefix, category),
+	})
+	for {
+		attrs, err := it.Next()
+		if err == iterator.Done {
+			break
+		}
+		if err != nil {
+			return nil, fmt.Errorf("Bucket(%q).Objects: %v", BucketName, err)
+		}
+		path := attrs.Name
+		if isFullPathToExample(path) {
+			infoPath := filepath.Join(path, MetaInfoName)
+			rc, err := bucket.Object(infoPath).NewReader(ctx)
+			if err != nil {
+				logger.Errorf("Object(%q).NewReader: %v", infoPath, err.Error())
+				continue
+			}
+			data, err := ioutil.ReadAll(rc)
+			if err != nil {
+				logger.Errorf("ioutil.ReadAll: %v", err.Error())
+				continue
+			}
+			exampleInfo := pb.Example{}
+			err = json.Unmarshal(data, &exampleInfo)
+			if err != nil {
+				logger.Errorf("json.Unmarshal: %v", err.Error())
+				continue
+			}
+
+			exampleInfo.CloudPath = path
+			exampleInfo.Name = filepath.Base(path)
+			cd.writeExample(path, &examples, &exampleInfo)
+			rc.Close()
+		}
+	}
+	return &examples, nil
+}
+
+func isFullPathToExample(path string) bool {
```

Comment:
suppose we have structure /SDK_JAVA/category1/example1. You calculate the count of /, so it's equal to 3 first condition is true. However, the second one is false. We expect / at the end of the path? Why? Based on the doc, we don't.  Can we add last / automatically if it is not present?

Reply:
Changed the structure description